Automatically change tab when saving resource and invalid values/states are present in another tab

OptimusCrime opened the pull request:
### What does it do?
Implement an extension of the current tab behavior when updating/creating a resource. If you happen to have a duplicated alias/uri and the processor returns an error, you will not automatically see the correct tab where the error is, if you are not already on that tab. See gif below.

**NOTE: This illustration is no longer valid. This PR now handles illegal states for all fields in the form, not only uri/alias.**

New behavior:
![alias-new-behavior](https://cloud.githubusercontent.com/assets/2326278/20855898/1c638212-b904-11e6-949b-251da02127b9.gif)

### Why is it needed?
Solves an annoying issue that most of us have experienced at some point.

### Related issue(s)/PR(s)
#13202

OptimusCrime committed:
Automatically change tab to settings if failure and somewhere else

Jako commented:
I think this is not enough (but it's nice for a start). A resource could be not saveable when a TV is required.

A field (should have an `x-form-field` class) with an error gets a `x-form-invalid` class. Searching `#modx-content` for those elements, looking if it is member of one or more tab content (`x-tab-panel`) and activating all tabs for the first element found should make it.

sdrenth commented:
Since this PR has been untouched for a while, is it perhaps an idea to accept this PR and maybe create another issue for the solution provided by @Jako? Or is @OptimusCrime planning on working on this PR any further?

I've tested the PR and I can confirm that it works.

OptimusCrime commented:
I was planning on trying to take a look at it. I just have way too much going on these days. If I find the time, I'll look at this first.

OptimusCrime committed:
Make automatic tab switcher handle any kind of invalid state

OptimusCrime commented:
I have expanded on the original PR. I implemented a new event which is fired upon failing to submit the form. This change had to be done to the `modx.component.js` file, as no events were available when a form contains an invalid state.

In the new event we are passed the entire MODX Resource component. We grab each tab form and simply loop the children recursively until we either hit the bottom, or we find an element with `x-form-invalid` class attached. Note that this approach is different from the validation done in `modx.component.js`. Their implementation uses the `validate()` method, which is -- as far as I can tell -- much more costly to call, especially on the entire form. I might be incorrect about this, however.

Some minor details about validating aliases exist, as this code was copied and pasted over in the new event handler.

I have tested the implementation, and it works like a charm!

pixelchutes commented on manager/assets/modext/widgets/resource/modx.panel.resource.js:
```diff
@@ -136,6 +137,7 @@ Ext.extend(MODx.panel.Resource,MODx.FormPanel,{
     }
 
     ,beforeSubmit: function(o) {
+        console.log('done');
```
Debug log? Should this be cleaned up?

sdrenth commented on manager/assets/modext/widgets/resource/modx.panel.resource.js:
```diff
@@ -200,6 +202,70 @@ Ext.extend(MODx.panel.Resource,MODx.FormPanel,{
+
+        // The next check needs the tabs index value to check if it is hidden or not
+        if (tabs && tabs.items && tabs.items.keys) {
+            tabs_index = tabs.items.keys.indexOf(tab_name)
```
Missing semicolon at end of line

OptimusCrime committed:
Fix mistakes, implement handling for duplicate alises

OptimusCrime commented:
Fixed comments, and implemented behavior for duplicate aliases.

See it in action below. A few notices: 

- Check and handling for duplicate aliases is only possible if the form itself was submitted. This means that any other mistakes (such as missing TVs) needs to be fixed first.
- If the TVs are categorized, this code does not switch to the correct category. This might be a future improvement.

![behavior_demo](https://user-images.githubusercontent.com/2326278/30037450-2199826e-91bb-11e7-856d-46bb8d5f5e38.gif)

I've removed the WIP tag. This should be good for merging. I could squash the commits it you'd like?

Mark-H commented:
Thanks @OptimusCrime
